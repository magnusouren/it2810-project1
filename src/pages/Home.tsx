/* eslint-disable @typescript-eslint/no-unused-vars */
import './Home.css';

import { useQuery } from '@tanstack/react-query';
import axios from 'axios';
import { FC } from 'react';

import { Category } from '../components/category/Category';
import { CategoryType } from '../types';

export const Home: FC = () => {
  console.log('Home');

  // tanstack query to get all categories
  const { data, isLoading, isSuccess } = useQuery<CategoryType[]>(['categories'], () =>
    axios
      .get('https://www.thecocktaildb.com/api/json/v1/1/list.php?c=list')
      .then((res) => res.data.drinks.map((drink: { strCategory: string }) => drink.strCategory)),
  );

  if (isLoading) return <div>Loading...</div>;

  return (
    <>
      <div className='wallpaper'>
        <h2>Categories:</h2>
      </div>
      {data && isSuccess && (
        <div className='categories'>
          {data.sort().map((category: CategoryType) => (
            <Category key={category} category={category} />
          ))}
        </div>
      )}
    </>
  );
};
