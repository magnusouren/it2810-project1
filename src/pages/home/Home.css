.categories {
  display: flex;
  flex-wrap: wrap;
  margin: 0 auto;
  width: 60%;
  gap: 1rem;
}

h2 {
  width: 60%;
  margin: 0 auto 1rem auto;
  padding-bottom: 0.5rem;
  border-bottom: solid 2px #f2c2a6;
}
