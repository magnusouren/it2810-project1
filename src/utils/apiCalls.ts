import axios from 'axios';

export const fetchCocktailById = async (id: string) => {
  const response = await axios.get(`https://www.thecocktaildb.com/api/json/v1/1/lookup.php?i=${id}`);
  return response.data;
};
