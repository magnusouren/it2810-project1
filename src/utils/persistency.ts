import { DrinkOfTheDay } from '../types';

const addFavorite = (id: string) => {
  const favorites = getFavorites();
  favorites.push(id);
  localStorage.setItem('drinkFavorites', JSON.stringify(favorites));
};

export const getFavorites = () => {
  const favorites = localStorage.getItem('drinkFavorites');
  if (favorites) {
    return JSON.parse(favorites);
  }
  return [];
};

const removeFavorite = (id: string) => {
  const favorites = getFavorites();
  const newFavorites = favorites.filter((favorite: string) => favorite !== id);
  localStorage.setItem('drinkFavorites', JSON.stringify(newFavorites));
};

export const toggleFavorite = (id: string) => {
  const favorites = getFavorites();
  if (favorites.includes(id)) {
    removeFavorite(id);
  } else {
    addFavorite(id);
  }
};

export const isFavorite = (id: string) => {
  const favorites = getFavorites();
  return favorites.includes(id);
};

export const getDrinkOfTheDay = (date: string) => {
  const drinkOfTheDay = localStorage.getItem(date);
  if (drinkOfTheDay) {
    return JSON.parse(drinkOfTheDay);
  }
  return null;
};

export const setDrinkOfTheDay = (date: string, drink: DrinkOfTheDay) => {
  localStorage.setItem(date, JSON.stringify(drink));
};
