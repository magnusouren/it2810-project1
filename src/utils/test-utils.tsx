import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

// Function to render a component wrapped inside a browserrouter for testing purposes
export const renderWithRouter = (component: JSX.Element) => render(<BrowserRouter>{component}</BrowserRouter>);

const queryClient = new QueryClient();

export const renderWithRouterAndQueryClient = (component: JSX.Element) =>
  render(
    <QueryClientProvider client={queryClient}>
      <BrowserRouter>{component}</BrowserRouter>
    </QueryClientProvider>,
  );
