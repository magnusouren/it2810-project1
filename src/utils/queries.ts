import axios, { isCancel } from 'axios';

import { Drink, DrinkOfTheDay, Ingredient } from '../types';
import { setDrinkOfTheDay } from './persistency';

export const fetchDrinkById = async (id?: string) => {
  if (!id) {
    return null;
  }
  const response = await axios
    .get(`https://www.thecocktaildb.com/api/json/v1/1/lookup.php?i=${id}`)
    .then((response) => {
      // Extract drink data
      const drinkData = response.data.drinks[0];
      const ingredients: Array<Ingredient> = [];
      const drink: Drink = {
        idDrink: drinkData.idDrink,
        strDrink: drinkData.strDrink,
        strGlass: drinkData.strGlass,
        ingredients: [],
        strInstructions: drinkData.strInstructions,
        strDrinkThumb: drinkData.strDrinkThumb,
        strCategory: drinkData.strCategory,
        strAlcoholic: drinkData.strAlcoholic,
      };

      // Extract ingredients
      for (let i = 1; i <= 15; i++) {
        const ingredient: string = drinkData[`strIngredient${i}`];
        const measure: string = drinkData[`strMeasure${i}`];
        if (ingredient) {
          ingredients.push({ ingredient, measure });
        } else {
          break;
        }
      }

      drink.ingredients = ingredients;
      return drink;
    })
    .catch((error) => {
      if (isCancel(error)) {
        console.log('Request cancelled', error.message);
        return null;
      }
      if (error instanceof TypeError) {
        return null;
      } else {
        console.error('Error fetching drink:', error);
        return null;
      }
    });
  return response;
};

export const fetchDrinkOfTheDay = async (currentDate: string) => {
  return axios
    .get('https://www.thecocktaildb.com/api/json/v1/1/random.php')
    .then((response) => {
      const randomDrink = response.data.drinks[0];

      const drink: DrinkOfTheDay = {
        drinkId: randomDrink.idDrink,
        strDrink: randomDrink.strDrink,
        strDrinkThumb: randomDrink.strDrinkThumb,
        strCategory: randomDrink.strCategory,
        strGlass: randomDrink.strGlass,
        strAlcoholic: randomDrink.strAlcoholic,
      };

      // Store the new random drink ID in local storage with the current date
      setDrinkOfTheDay(currentDate, drink);

      return drink;
    })
    .catch((error) => {
      console.error('Error fetching drink:', error);
      return null;
    });
};
