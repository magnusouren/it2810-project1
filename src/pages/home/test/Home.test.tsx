import { screen, waitFor } from '@testing-library/react';

import { renderWithRouterAndQueryClient } from '../../../utils/test-utils';
import { Home } from '../Home';

describe('Accordion test', () => {
  it('Should match snapshot', async () => {
    const { container } = renderWithRouterAndQueryClient(<Home />);
    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });

  it('Should show title and links', async () => {
    const { container } = renderWithRouterAndQueryClient(<Home />);
    await waitFor(() => {
      expect(container.innerHTML).toContain('Beer');
    });
    expect(screen.getByText('Beer')).toBeTruthy();
  });
});
