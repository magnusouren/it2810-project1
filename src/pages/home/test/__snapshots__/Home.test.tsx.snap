// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Accordion test > Should match snapshot 1`] = `
<div>
  <div>
    Loading...
  </div>
</div>
`;
